Antlr tokenizer

mpherman2 opened the pull request:
Make Javascript minimization happen using a antlr-grammar tokenizer 

Everything in the src/python/bot/minimizer/grammars directory is auto-generated by antlr with the exception of JavaScriptBaseLexer.py which was hand translated from a Java file at : https://github.com/antlr/grammars-v4/blob/master/javascript/Java/JavaScriptBaseLexer.java

Dor1s commented on src/python/bot/tasks/minimize_task.py:
```diff
@@ -1395,13 +1419,14 @@ def minimize_file(file_path,
   if file_path.endswith('.js'):
-    return do_js_minimization(test_function, get_temp_file, data, deadline,
-                              threads, cleanup_interval, delete_temp_files)
+    js_tokenizer = AntlrTokenizer(JavaScriptLexer)
```
in this current form the change will unconditionally switch the minimization for JS files to the antlr-based tokenizer, but I guess we don't want it right now. Will need to decide whether we make this behavior depend on some additional parameter, or disable for now. Not an issue you can address right away, but we'll need to think before landing this

Dor1s replied:
@mpherman2 do you know what was @mbarbella-chromium 's plan on this? I don't think we should switch to the new minimizer until we're happy with its performance

mpherman2 replied:
Not sure, he did not say. I think either setting a parameter or just disabling it entirely for now are both fine. What do you think?

Dor1s replied:
I would disable it for now. Enabling it is fast and easy.

mpherman2 replied:
Sounds good

mpherman2 commented on src/python/bot/minimizer/grammars/JavaScriptLexer.interp:
```diff
@@ -0,0 +1,413 @@
+token literal names:
```
I'm not sure. This was an auto-generated file from antlr, so I did not change anything

mpherman2 commented on src/python/tests/core/bot/minimizer/antlr_tokenizer_test.py:
```diff
@@ -0,0 +1,66 @@
+from bot.minimizer.grammars.JavaScriptLexer import JavaScriptLexer
+
+class AntlrTokenizerTest(unittest.TestCase):
+  '''Tests for AntlrTokenizer'''
```
Done

mpherman2 commented on src/python/tests/core/bot/minimizer/antlr_tokenizer_test.py:
```diff
@@ -0,0 +1,66 @@
+          }"""
+
+    tokens = tokenizer.tokenize(txt)
+    self.assertEqual(tokens, ['async', ' ', 'function', ' ', 'process', '(',
```
Done

mpherman2 commented on src/python/bot/minimizer/grammars/JavaScriptBaseLexer.py:
```diff
@@ -0,0 +1,87 @@
+  def ProcessStringLiteral(self):
+    if self._lastToken == None or self._lastToken.type == self.OpenBrace:
+      text = super(JavaScriptBaseLexer, self).text
+      if text == "\"use strict\"" or text == "'use strict'":
```
Done

mpherman2 commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,43 @@
+# limitations under the License.
+"""Antlr Tokenizer"""
+
+from antlr4 import *
```
Done

mpherman2 commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,43 @@
+from antlr4 import *
+
+class AntlrTokenizer():
+  '''Tokenizer. Takes an Antlr Lexer created using
```
Done

mpherman2 commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,43 @@
+  def __init__(self, lexer):
+    self._lexer = lexer
+
+  def fill(self, stream):
```
Done

mpherman2 commented on src/python/bot/tasks/minimize_task.py:
```diff
@@ -1348,6 +1349,29 @@ def do_libfuzzer_cleanse(testcase, testcase_file_path, expected_crash_state):
   return output_file_path
 
 
+def do_antlr_tokenized_minimization(test_function, get_temp_file, data, deadline, threads,
```
Done

mpherman2 commented:
Addressed some comments, please take another look.

Dor1s commented:
+Oliver to take a look while Marty is out, I'll also do another pass today

Dor1s commented:
Since linter complains about auto-generated files, I'd suggest the following:

1) add an extra condition to https://github.com/google/clusterfuzz/blob/d709ba572677c743dc838438e319fb1186f3d3ac/src/local/butler/lint.py#L188 to skip files in the `grammars` dir, I'd probably create a small helper function e.g. `is_auto_generated_file` and move the existing check over there + add another check using `os.path.dirname(f)`

2) move any manually-written files out of that directory

3) add a `README.md` file clarifying that all files in that directory are auto-generated and therefore exempt for linting, "do not add files written by people", or something like that

Dor1s commented:
> Since linter complains about auto-generated files, I'd suggest the following:

Sorry I didn't realize previously that linter would not let us to commit these files, when I said not to worry about linter errors in the auto-generated files. I mean, we still don't need to worry about the errors themselves, but we have to tell the linter that it's ok and it should not complain. Sorry if I confused you, @mpherman2  :)

Dor1s commented on src/local/requirements.txt:
```diff
@@ -8,3 +8,4 @@ paramiko==2.6.0
 protobuf==3.6.1
 psutil==5.4.7
 tensorflow==1.8.0
+antlr4-python2-runtime==4.7.2
```
I think this also needs to be added to `docker/base/requirements.txt`

mpherman2 replied:
Done

Dor1s commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,48 @@
+    end = self.fill(stream)
+
+    tokens = stream.getTokens(0, end)
+    return [token.text for token in tokens]
```
do we need to return a list, or would `yield` work here?

mpherman2 replied:
I'm pretty sure it has to be a list. The entire list of tokens is an attribute for the testcase passed to the minimizer.

Dor1s commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,48 @@
+    tokens = stream.getTokens(0, end)
+    return [token.text for token in tokens]
+
+  def combine(self, tokens):
```
is this function really needed? if it's a part of the interface, please add a comment

mpherman2 replied:
The default token combiner is to join on new line which is not the correct behavior. This seemed to be the best place to put the combiner to show that its used for the antlr grammar. It needs to be a method because the method is passed as a parameter to the minimizer. I'll add a comment.

Dor1s commented on src/python/bot/minimizer/grammars/JavaScriptBaseLexer.py:
```diff
@@ -0,0 +1,82 @@
+# limitations under the License.
+"""Translation from Java code for JavaScriptBaseLexer made to work with
+JavaScriptLexer"""
+
```
Is this also an auto-generated file, or did you accidentally miss my comments below?

Dor1s replied:
Maybe you forgot to upload the changes you made locally :)

mpherman2 replied:
Sorry. Messed the comments from before. This is not an auto-generated file. The JavaScriptLexer is an auto-generated file, but it was generated to work with this JavaScriptBaseLexer file which was not available in python. I had to translate it from Java. I can change the import, but the fact that the naming conventions unusual I can't change without changing the auto generated file. Even though this file isn't necessary auto-generated, I'm going to leave it with the rest of the code not being linted.

mpherman2 commented on src/python/bot/minimizer/grammars/JavaScriptBaseLexer.py:
```diff
@@ -0,0 +1,87 @@
+
+from antlr4 import *
+
+class JavaScriptBaseLexer(Lexer):
```
See above, but I'm also not sure as this was translated. I will clarify more in the top level comment the origin of this file and what it is doing

mpherman2 commented on src/python/bot/minimizer/grammars/JavaScriptBaseLexer.py:
```diff
@@ -0,0 +1,87 @@
+    self._useStrictDefault = False
+    self._useStrictCurrent = False
+
+  def IsStartOfFile(self):
```
That is the naming convention that the auto-generated file uses.

Dor1s replied:
weird, thanks!

mpherman2 commented on src/python/bot/minimizer/grammars/JavaScriptBaseLexer.py:
```diff
@@ -0,0 +1,87 @@
+"""Translation from Java code for JavaScriptBaseLexer made to work with
+JavaScriptLexer"""
+
+from antlr4 import *
```
Done

mpherman2 commented:
Addressed some of the comments above. I'm not sure how to fix the two errors I'm getting on the pull request though. I would appreciate if you took a look. Thanks

mpherman2 commented:
/gcbrun

inferno-chromium commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,49 @@
+  and allows user to tokenize files using that grammar"""
+
+  def __init__(self, lexer):
+    """constructor"""
```
Can remove this.

mpherman2 replied:
Done

inferno-chromium commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,49 @@
+
+  def fill(self, stream):
+    """Helper function. antlr4.CommonTokenStream.fill should work, but
+    it does not fetch all of the tokens. This is a replacement that works"""
```
nit:end with period.

mpherman2 replied:
Done

inferno-chromium commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,49 @@
+    i = 0
+    while stream.fetch(1):
+      i += 1
+    return i
```
Shouldn't this return the tokens itself and not just the count, then you shouldn't need to do stream.getTokens

mpherman2 replied:
fetch(x) fills a buffer with x tokens, and stream.getTokens retrieves them from the buffer. Not sure why it works this way either.

inferno-chromium replied:
Ok fine then.

inferno-chromium commented on src/python/bot/minimizer/grammars/JavaScriptBaseLexer.py:
```diff
@@ -0,0 +1,82 @@
+# Copyright 2019 Google LLC
```
should these files be in bot/tokenizer/grammars ?

mpherman2 replied:
This is where we have put everything related to antlr grammars. This includes .g4 files as well as the Lexers and Parsers generated and used by antlr. Do you think the files should be split up? Or maybe the directory just needs to be renamed?

inferno-chromium replied:
i mean instead of minimizer directory, use tokenizer root directory. everything inside a grammars sub-dir is fine.

mpherman2 replied:
I misread that, sorry! I agree, the grammars should be in the same dir as the tokenizer.

inferno-chromium commented on src/python/bot/tasks/minimize_task.py:
```diff
@@ -37,6 +37,8 @@
 from bot.minimizer import js_minimizer
 from bot.minimizer import js_tokenizer
 from bot.minimizer import minimizer
+#from bot.minimizer.antlr_tokenizer import AntlrTokenizer
```
Can skip this step. Lets not use the code yet in minimizer, this part can go in next CL, also we don't check in commented code.

mpherman2 replied:
Done! It makes sense not to check in commented code. Thanks for the explanation

Dor1s commented on src/local/butler/lint.py:
```diff
@@ -167,6 +167,12 @@ def yaml_validate(file_path):
     _error('Failed: Invalid yaml file %s.\n\n%s' % (file_path, e))
 
 
+def auto_generated(f):
```
nit: and maybe change `f` to `path` or `filepath`

mpherman2 replied:
Done

Dor1s commented on src/local/butler/lint.py:
```diff
@@ -185,7 +191,7 @@ def execute(_):
       f for f in file_paths if f.endswith('.py') and
       # Exclude auto-generated files.
-      not f.endswith('_pb2.py') and not f.endswith('_pb2_grpc.py')
+      not auto_generated(f)
```
nit: can remove the comment on line 193 as the function name is self-descriptive `is_auto_generated`

mpherman2 replied:
Done

Dor1s commented on src/python/bot/tasks/minimize_task.py:
```diff
@@ -1395,8 +1422,14 @@ def minimize_file(file_path,
   if file_path.endswith('.js'):
-    return do_js_minimization(test_function, get_temp_file, data, deadline,
-                              threads, cleanup_interval, delete_temp_files)
+    #New Antlr-Tokenized minimization disabled until happy with behavior
```
+1 to the Abhishek's comment above, just delete all these commented lines for now

mpherman2 replied:
Done

mpherman2 commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,49 @@
+# Copyright 2019 Google LLC
```
Currently the js tokenizer and html tokenizer live in the minimizer dir. Should we move all of these? Where else do we use the tokenizers besides minimization?

mpherman2 commented on src/local/butler/lint.py:
```diff
@@ -167,6 +167,12 @@ def yaml_validate(file_path):
+def auto_generated(f):
+  """Check if file is auto-generated so we dont lint it"""
+  return f.endswith('_pb2.py') or f.endswith('pb2_grpc.py') or \
+         os.path.dirname(f) == "src/python/bot/minimizer/grammars"
```
Done

mpherman2 commented on src/python/bot/minimizer/antlr_tokenizer.py:
```diff
@@ -0,0 +1,49 @@
+class AntlrTokenizer(object):
+  """Tokenizer. Takes an Antlr Lexer created using
+  $ antlr4 -Dlanguage=Pythonn <AntlrGrammar.g4>
+  and allows user to tokenize files using that grammar"""
```
Done

mpherman2 commented:
/gcbrun

inferno-chromium commented on src/python/bot/tasks/minimize_task.py:
```diff
@@ -1395,8 +1420,8 @@ def minimize_file(file_path,
   if file_path.endswith('.js'):
-    return do_js_minimization(test_function, get_temp_file, data, deadline,
-                              threads, cleanup_interval, delete_temp_files)
+    do_js_minimization(test_function, get_temp_file, data, deadline, threads,
```
why remove the return ?

mpherman2 replied:
Good catch, that was a mistake. When we update minimization in the next CL we should add more extensive unit tests. I don't think this code branch is tested at all right now.

inferno-chromium commented on src/python/bot/tasks/minimize_task.py:
```diff
@@ -1348,6 +1348,31 @@ def do_libfuzzer_cleanse(testcase, testcase_file_path, expected_crash_state):
   return output_file_path
 
 
+def do_antlr_tokenized_minimization(test_function, get_temp_file, data,
```
This function is unused and should be added in next cl.

mpherman2 replied:
Done

mpherman2 commented:
/gcbrun
